Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java

```diff
@@ -304,18 +309,31 @@ public String getJobUuid() {
                                         return Observable.from(logLines)
                                                 .map(line -> new AbstractMap.SimpleImmutableEntry<>(Log, line));
                                     } else {
-                                        // If logs we get is empty, we send an onError() message,
-                                        // which will get into RetryWhen and then retry with limited times
-                                        String errorMsg = "Livy submission log is empty.";
-                                        log().warn(errorMsg);
-                                        ctrlInfo(errorMsg);
-                                        return Observable.error(new SparkJobException(errorMsg));
+                                        // If logs we get is empty, we will check the job state first.
+                                        // If job is not in STARTING state, we will stop getting submission log,
+                                        // or else we will retry getting submission log again.
+                                        return getResponsePayloadWithState()
+                                                .flatMap(responsePayload -> {
+                                                    // FIXME:
+                                                    // For HDI spark job, the ending condition is `jobState != starting || appIdIsAllocated`
+                                                    // However, currently the response has no appId all the time so we ignored the app id check
+                                                    if (!responsePayload.getState().equalsIgnoreCase(SparkBatchJobState.STARTING.toString())) {
+                                                        String errorMsg = "Spark job is in not in STARTING state, so we should stop getting submission log.";
+                                                        log().warn(errorMsg);
+                                                        ctrlInfo(errorMsg);
+                                                        return Observable.error(new SparkSubmissionLogFinishedException(errorMsg));
```

Comment:
May return `Observable.empty()` would be better than throw exceptions.

Reply:
Move logs empty check and exception throwing code to `repeatWhen`